### ISSUE TITLE
with-polyfills: Dynamic import of polyfill causes chunk load error

### ISSUE THREAD

**Issue by jnv** [vercel/next.js#8189]
# Examples bug report

## Example name

[with-polyfills](https://github.com/zeit/next.js/tree/master/examples/with-polyfills)

## Describe the bug

`with-polyfills` shows how to polyfill missing features prior to initialization of Next app. However, I would like to load polyfills only for the missing features using the dynamic import.

For example IntersectionObserver is supported widely in the latest versions of browsers, so it's usually not necessary to load the [intersection-observer polyfill](https://github.com/w3c/IntersectionObserver/tree/master/polyfill). Thus I would like to import the polyfill only when the feature is missing like this:

```js
// client/polyfills.js

async function loadPolyfills() {
  if (typeof window.IntersectionObserver === 'undefined') {
    await import('intersection-observer')
  }
}
export default loadPolyfills()
```

When I do that the application crashes when browser attempts to load the polyfill with the following error:

```
Unhandled Rejection (ChunkLoadError): Loading chunk 0 failed.
(error: http://localhost:3000/static/chunks/0.js)
```

Apparently Webpack points to the chunk in a wrong location, since the chunk is available in `_next` directory, i.e. `http://localhost:3000/_next/static/chunks/0.js`

## To Reproduce

I have prepared a repository with reproduction of the issue. The feature detection is [commented out](https://github.com/jnv/next-with-polyfills-dynamic-import/blob/master/client/polyfills.js) to reproduce the behavior in any browser.

1. Download or clone https://github.com/jnv/next-with-polyfills-dynamic-import
2. Install dependencies with `yarn`
3. Start application with `yarn dev`
4. Go to `http://localhost:3000/`
5. See the error

## Expected behavior

Browser loads the chunk without error, before Next initialization.

## Screenshots

![Error in browser](https://user-images.githubusercontent.com/616767/62223424-3029a980-b3b5-11e9-9a6d-ef9d9d926270.png)

## System information

- OS: macOS
- Browser: does not apply
- Version of Next.js: 9.0.2

**Comment by JackCA** [comment]
Has anyone found a workaround for this?

**Comment by jnv** [comment]
Yeah, I also wonder about that. We've ended up sourcing the polyfill from https://polyfill.io/ which is quite suboptimal.

**Comment by JackCA** [comment]
Yeah, not ideal. I just stopped dynamically importing it, as it was only 2KB gzipped.

I wonder if moving the dynamic import into another file that wraps it would help next.js resolve it better

**Comment by nickrttn** [comment]
I'm seeing the same thing. Using the same configuration as `with-polyfills` example, I noticed that dynamic `import()` tries to load the script I added from a different directorypath than other entries, causing a 404.

**Loading the polyfills chunk**

![Screenshot 2019-12-02 at 10 54 51](https://user-images.githubusercontent.com/2504906/69949666-3e6a0c80-14f2-11ea-89de-ef5683bd135a.png)


**Loading another chunk**

![Screenshot 2019-12-02 at 10 54 39](https://user-images.githubusercontent.com/2504906/69949712-504baf80-14f2-11ea-8b03-43b21c20ef64.png)

You can see that in the first screenshot, the `/_next/` prefix is missing from the path.

**Comment by jnv** [comment]
I suspect that Next does some extra stuff with paths mapping either during or after the build and custom bundle entry point doesn't get transformed.

**Comment by mulholio** [comment]
This is quite a blocker for what I'm working on so I'm going to dive into the codebase and see if I can suggest a fix. Any time-saving recommendations of where to look appreciated!

**Comment by jnv** [comment]
@mulholio I've tried to find some leads in the codebase (e.g. references to `_next/static/chunks` directory etc.), but found only a few integration tests. Personally I'd start with [build](https://github.com/zeit/next.js/tree/canary/packages/next/build) directory in the `next` package, and especially [webpack-config](https://github.com/zeit/next.js/blob/canary/packages/next/build/webpack-config.ts). Creating an integration test from the example may be also helpful once you identify the source of this issue.

**Comment by mulholio** [comment]
I think this is actually something to do with Next Server which is what handles the _next prefix ([source](
https://github.com/zeit/next.js/issues/8158#issuecomment-516010976))

**Comment by mulholio** [comment]
This issue is of a very similar issue but no mention of what resolved it: https://github.com/zeit/next.js/issues/4993

**Comment by mulholio** [comment]
I've given up on this now but still very keen to see if anyone else comes up with a solution (anyone @zeit???)

**Comment by JackCA** [comment]
I'd remove the word "polyfill" from this issue because it seems to just flat out happen for dynamic imports. I am still continually hit by this

**Comment by jnv** [comment]
@JackCA Do you have any other examples where this happens? I ran into this specifically with modified Webpack config to run dynamic imports before Next.js initialization.

But when I think of it, I don't think we use import function anywhere outside of `next/dynamic`.

**Comment by JackCA** [comment]
This is happening for me when using a raw dynamic import of a library I don't want/need on the initial page load.

This doesn't happen to me for `next/dynamic` but as this isn't a React component that's being dynamically imported, I don't believe I can use that functionality.

Some additional context is that I'm seeing evidence that this is a "timeout" of some sort. [Here is a public Sentry.io issue](https://sentry.io/share/issue/0ed791a1ab0b473ab2f8352e47cf3d13/) that may help.

**Comment by gfsd3v** [comment]
having the same issue, anybody have some info about this ?

**Comment by jnv** [comment]
FWIW I think the proposed plugins system could solve this if we can hook into the lifecycle before hydratation: https://github.com/vercel/next.js/discussions/9133

**Comment by voronianski** [comment]
experiencing the same issue, any estimate on this?

**Comment by ccambournac** [comment]
Hi,

I experienced the same issue and fixed it inside [custom Webpack configuration](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config) file by adding the following:
```
// One-time procedures to run
if (!isServer) {
	// For polyfills chunks to be correctly loaded
	config.output.publicPath = '/_next/';
}
```

**Comment by voronianski** [comment]
It's possible to fix chunk error with this line inside `next.config.js` -

```js
if (!isServer) {
  config.output.publicPath = '/_next/';
}
```

however the file in polyfills.js that's asynchronously imported is loaded after the actual file (e.g. commons.js) that requires the polyfill.

**Comment by voronianski** [comment]
did anyone had a chance to make Next.js work with loading polyfills asynchronously only when they are needed? I'm thinking of doing it in App.js instead 🤔

Here is the example of my `polyfills.js`. As I said it loads the polyfill file asynchrnously and after the actual code that needs the polyfill - 

```js
import isServer from '~/util/is-server';

async function loadIntersectionObserverPolyfill() {
  const supportsIntersectionObserver =
    'IntersectionObserver' in window &&
    'IntersectionObserverEntry' in window &&
    'intersectionRatio' in window.IntersectionObserverEntry.prototype;

  if (!supportsIntersectionObserver) {
    await import(
      // eslint-disable-next-line max-len
      /* webpackChunkName: "intersection-observer-polyfill" */ 'intersection-observer'
    );
  }
}

const Polyfills = (async () => {
  if (!isServer) {
    await loadIntersectionObserverPolyfill();

    // will work but will load polyfill code everywhere
    // require('intersection-observer');
  }
})();

export default Polyfills;
```

I assume the ultimate solution for this problem would be somehow waiting until `polyfills.js` is loaded only after start loading other chunks (somewhere [here](https://github.com/vercel/next.js/blob/master/packages/next/pages/_document.tsx#L776)).

**Comment by boschni** [comment]
Another possibility would be to create a custom `_document` and conditionally append polyfill scripts with `document.write`:
```js
export default class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head>
          <script
            dangerouslySetInnerHTML={{
              __html: `
              if (!Intl.DisplayNames) {
                document.write('<script src="/intl-displaynames-' + document.documentElement.lang + '.hash.js"></scr' + 'ipt>');
              }`,
            }}
          />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}
```
But it would be great if there was a better way to execute some asynchronous code before Next.js initializes.

**Comment by jnv** [comment]
Maybe the recently introduced [script component](https://nextjs.org/docs/basic-features/script) with beforeInteractive can help here, but it seems to be rather designed for importing 3rd party scripts.

**Comment by CamPCatDev** [comment]
need help?

**Comment by CamPCatDev** [comment]
need help?

**Comment by jnv** [comment]
Feel free to take over this issue, I no longer work on the project where this was relevant.

**Comment by balazsorban44** [comment]
This issue has been automatically closed because it wasn't verified against next@canary. If you think it was closed by accident, please leave a comment. If you are running into a similar issue, please open a new issue with a reproduction. Thank you.
